File: lc/coreprep/implementation/patterns_reference_implementation/merge_intervals/AllConflictingAppointments.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n2) / O(1)
public class AllConflictingAppointments {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  static class Pair {
    Interval first;
    Interval second;

    Pair(Interval first, Interval second) {
      this.first = first;
      this.second = second;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int s = in.nextInt();
          int e = in.nextInt();
          a.add(new Interval(s, e));
        }

        BiFunction<Interval, Interval, Boolean> isOverlapping =
            (Interval x, Interval y) -> {
          return (x.start >= y.start && x.start < y.end) ||
              (y.start >= x.start && y.start < x.end);
        };

        List<Pair> res = new ArrayList<>();
        for (int i = 0; i < n; i++) {
          for (int j = i + 1; j < n; j++) {
            if (isOverlapping.apply(a.get(i), a.get(j))) {
              res.add(new Pair(a.get(i), a.get(j)));
            }
          }
        }

        System.out.println(res.size());
        for (var p : res) {
          System.out.printf("%d %d %d %d\n", p.first.start, p.first.end,
                            p.second.start, p.second.end);
        }
      }
    }
  }
}

File: lc/coreprep/implementation/patterns_reference_implementation/merge_intervals/all_conflicting_appointments.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval(int start, int end) : start(start), end(end) {}
};

vector<pair<Interval, Interval>> all_conflicts(const vector<Interval> &a) {
  const auto has_conflict = [](const Interval &a, const Interval &b) {
    return (a.start >= b.start && a.start < b.end) ||
           (b.start >= a.start && b.start < a.end);
  };

  vector<pair<Interval, Interval>> res;
  for (int i = 0; i < a.size(); i++) {
    for (int j = i + 1; j < a.size(); j++) {
      if (has_conflict(a[i], a[j])) {
        res.emplace_back(make_pair(a[i], a[j]));
      }
    }
  }

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a;
    int s, e;
    for (int i = 0; i < n; i++) {
      cin >> s >> e;
      a.emplace_back(Interval(s, e));
    }

    auto res = all_conflicts(a);
    cout << res.size() << "\n";
    for (auto r : res) {
      cout << r.first.start << " " << r.first.end << " " << r.second.start
           << " " << r.second.end << "\n";
    }
  }

  return 0;
}